### PR TITLE
fix-diag to sparse conversion

### PR DESCRIPTION
**Pull request by GiggleLiu**

### CODE REVIEW

**Comment by Roger-luo** [comment]
CI is not pass.

File: src/conversions.jl [hunk 1]
```diff
@@ -19,6 +19,11 @@ function SparseMatrixCSC(M::PermMatrix)
     SparseMatrixCSC(n, n, collect(1:n+1), order, M.vals[order])
 end
 
+function SparseMatrixCSC(M::Diagonal)
+    n = size(M, 1)
+    SparseMatrixCSC(n, n, collect(1:n+1), collect(1:n), M.diag)
```
Comment (Roger-luo) [inline src/conversions.jl]: seems this line is using the wrong args.

Comment (GiggleLiu) [reply src/conversions.jl]: What is the correct one?

**Comment by Roger-luo** [comment]
XRef: https://github.com/JuliaLang/julia/pull/32466
issue: https://github.com/JuliaLang/julia/issues/31770

I'll patch it here for now. But this PR should be in the coming release, so should be removed after 1.3 is out this year, or we would contain type piracy.

This issue is because the conversion was defined in `sparse` instead of `SparseMatrixCSC`, not just because the lack of constructors. 

https://github.com/JuliaLang/julia/blob/55e36cc308b66d3472990a06b2797f9f9154ea0a/stdlib/SparseArrays/src/sparsematrix.jl#L778

I don't think it would appear in v1.2, but should be in v1.3 period, not sure if this would get backported, and I think you missed the type conversion of the diagonal part,
it should be copied instead of view (`Diagonal` is not sub array, should use an explicit type for diagonal view to avoid heap allocation or use `UnsafeArray`).

```julia
function sparse(D::Diagonal{T}) where T
    m = length(D.diag)
    return SparseMatrixCSC(m, m, Vector(1:(m+1)), Vector(1:m), Vector{T}(D.diag))
end
```

**Comment by Roger-luo** [comment]
PS. I think this is one reason you want to get rid of this kind of auto conversion interface like `sparse`, the same as `mat` we should use explicit constructors more in the future.

**Pull request merged**